### ISSUE TITLE
Username is inconsistent for a blog in a folder

### ISSUE THREAD

**Issue by kvibber** [jdormit/pterotype#10]
After testing for a few weeks on https://blog.keysmash.xyz I've set up Pterotype on my long-running blog at https://www.hyperborea.org/journal/

It's sort of working, but the username is inconsistent because the blog is in a folder instead of at the top level of the website.

- Searching in Mastodon for blog@www.hyperborea.org doesn't find anything.
- Searching for blog@www.hyperborea.org/journal does turn up the profile...but only the first time I search. The profile removes the forward slash, leaving it as blog@www.hyperborea.orgjournal.
- Searching for blog@www.hyperborea.orgjournal on a new Mastodon instance doesn't work, and I think that's correct, because if it's never seen the username before it will be looking for something in the orgjournal TLD, which of course doesn't exist.
- Searching for blog@www.hyperborea.orgjournal after having searched for blog@www.hyperborea.org/journal on an instance *does* turn up the profile, and it can be followed.

So it's in a situation where depending on whether someone has searched for the profile or not, the username you need to look for to follow the site is different!

I imagine it's going to need some sort of redirect at the top level to ensure a username can be discovered at @blog@www.hyperborea.org or @blog-journal@www.hyperborea.org - something with the actual hostname, so it can be discovered without getting mangled.

This is using WordPress 4.9.8 and Pterotype 1.2.11.

Thanks!

**Comment by pfefferle** [comment]
See https://github.com/pterotype-project/pterotype/issues/9#issuecomment-440299905 and following comments

**Comment by kvibber** [comment]
@pfefferle Thanks. I was able to use the following rewrite rules to get blog@www.hyperborea.org to resolve using webfinger.net

```
RewriteCond ${unescape:%{QUERY_STRING}} resource=acct:(.+)
RewriteRule ^.well-known/webfinger /journal/index.php?well-known=webfinger&resource=acct:${unescape:%1}/journal [L]
```

On a new Mastodon instance, I'm able to find the profile by searching for @blog@www.hyperborea.org, and the profile that results is @blog@www.hyperborea.org/journal

On a Mastodon instance that already knows about @blog@www.hyperborea.org/journal, searching for @blog@www.hyperborea.org turns up a 422 error saying that the user already exists.

So I've still got a discovery problem where the "real" profile doesn't match the searchable profile the first time someone on a server looks for it, but the second time it's something else.

I imagine the solution is probably going to be reverting the change that adds the subfolder to the username, and generating a rewrite rule that the user can paste into their top-level config (assuming Apache, anyway)

**Comment by jdormit** [comment]
> I imagine the solution is probably going to be reverting the change that adds the subfolder to the username, and generating a rewrite rule that the user can paste into their top-level config (assuming Apache, anyway)

Yeah, I think that's exactly it. I'll see if I can get this done in the next couple of days.

**Comment by pfefferle** [comment]
@jdormit yes, I would highly recommend to remove the `path` from the `acct`.

This was already discussed on the mastodon repo a year ago https://github.com/tootsuite/mastodon/issues/752

**Comment by jdormit** [comment]
Okay, 1.2.12 reverts the earlier change and requires that `.well-known` exist at the root domain. It also removes the path from account paths, so @kvibber's actor account will be `@blog@www.hyperborea.org`. That means the correct rewrite rule should be:
```
RewriteCond ${unescape:%{QUERY_STRING}} resource=acct:(.+)
RewriteRule ^.well-known/webfinger /journal/index.php?well-known=webfinger&resource=acct:${unescape:%1} [L]
```

**Comment by kvibber** [comment]
Thanks! So, that *mostly* works. On Photog.social, everything seems to be working as expected. I can follow the blog at @blog@www.hyperborea.org, boost and reply to comments there. But at wandering.shop and booktoot.club, where I'd previously looked up @blog@www.hyperborea.org/journal, it still sees posts coming through on @blog@www.hyperborea.orgjournal - including the one I boosted on photog.social!

I can *see* the new actor name on wandering.shop and booktoot.club, but I can't follow it, and all the activity still shows up on the old actor name. I expect any new installations will be fine on this issue, but I'm not sure if there's a way I can clean it up or not.

**Comment by pfefferle** [comment]
Mastodon has a really strong caching... I also had this problem working on an ActivityPub plugin... I do not there is a solution except waiting...

**Comment by kvibber** [comment]
Yeah, I figured that was entirely on Mastodon's end.

On the plus side, interacting with it does seem to work either way. I'm able to reply to the post from both wandering.shop (which sees the old name) and photog.social (which sees the new name) and have the replies sync as comments.

**Comment by lapineige** [comment]
Hello,

How can I adapt this rewrite rule to Nginx ?

**Comment by lapineige** [comment]
I tried with
`if ($arg_resource ~ ^(?<account>acct:.+)) {
rewrite ^.well-known/webfinger /journal/index.php?well-known=webfinger&resource=$account last;
}`
(for Nginx)

But it still doesn't work.
Just to check, I try to reach `mydomain.tld/subpath/.well-known/webfinger` with my browser, I have this message (which seems normal, at least it shows that's working ?) : Expected a "resource" parameter

**Comment by lapineige** [comment]
:up: :)

**Comment by jdormit** [comment]
:tada: nice! What did you end up using for your rewrite rule?

**Comment by lapineige** [comment]
In fact I still don't know how to configure it properly, that's why I'm doing a little :up: :)

**Comment by lapineige** [comment]
:up: :)

**Comment by eldelacajita** [comment]
> That means the correct rewrite rule should be:
> 
> ```
> RewriteCond ${unescape:%{QUERY_STRING}} resource=acct:(.+)
> RewriteRule ^.well-known/webfinger /journal/index.php?well-known=webfinger&resource=acct:${unescape:%1} [L]
> ```

So if I understood correctly, if my blog is under `domain.com/blog`, y should use put those lines in .htaccess, changing `/journal` with `/blog`. Is that correct?

I'm not very familiar with .htaccess. Is it OK to just add the lines at the end, like this? 
```
# BEGIN WordPress
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteBase /blog/
RewriteRule ^index\.php$ - [L]
RewriteCond %{REQUEST_FILENAME} !-f
RewriteCond %{REQUEST_FILENAME} !-d
RewriteRule . /blog/index.php [L]
</IfModule>

# END WordPress

RewriteCond ${unescape:%{QUERY_STRING}} resource=acct:(.+)
RewriteRule ^.well-known/webfinger /blog/index.php?well-known=webfinger&resource=acct:${unescape:%1} [L]

```

**Comment by jdormit** [comment]
@eldelacajita you should put the new rewrite rules before the WordPress-generated ones, otherwise those will rewrite the URL before it has the chance to match the Webfinger rule

**Comment by lapineige** [comment]
Thanks for pointing that.

I did the change (with Nginx config), but it's still not working.

BTW do I have to install WebFinger plugin ?

**Comment by pfefferle** [comment]
@eldelacajita you should create a `.htaccess` file in the root of your webspace (not inside `/blog`) and add only the mentioned lines.

**Comment by pfefferle** [comment]
@eldelacajita if you are using the ActivityPub plugin, you have to also use a different endpoint `/blog/wp-json/activitypub/1.0/webfinger?resource=acct:${unescape:%1}`

**Comment by eldelacajita** [comment]
Many thanks, @pfefferle and @jdormit! I'll try that and let you know how it goes.
